lib/automation.go

```diff
@@ -126,3 +147,183 @@ func (automationImpl) Apply(scope scope, p *ApplyParams) (*ApplyResult, error) {
 	res.RunID = runID
 	return res, nil
 }
+
+// Deploy adds or updates a Dataset, creates or updates an associated Workflow, and, if deployParams.Apply is true, immediately runs the Workflow
+func (automationImpl) Deploy(scope scope, p *DeployParams) error {
+	log.Debugw("deploy", "dataset name", p.Dataset.Name, "peername", p.Dataset.Peername, "workflow id", p.Workflow.ID)
+	if p.Workflow.ID != "" {
+		wf, err := scope.Automation().GetWorkflow(p.Workflow.ID)
+		if err != nil {
+			return nil
+		}
+		if p.Workflow.DatasetID != wf.DatasetID {
+			return fmt.Errorf("deploy: given workflow and workflow on record have different DatasetIDs")
+		}
+		if p.Workflow.OwnerID != wf.OwnerID {
+			return fmt.Errorf("deploy: given workflow and workflow on record have different OwnerIDs")
+		}
+	}
+	go deploy(scope, p)
+	return nil
+}
+
+func deploy(scope scope, p *DeployParams) {
+	vi := dsref.ConvertDatasetToVersionInfo(p.Dataset)
+	ref := vi.SimpleRef().String()
+	deployPayload := event.DeployEvent{
+		Ref:        ref,
+		DatasetID:  p.Dataset.ID,
+		WorkflowID: p.Workflow.ID.String(),
+	}
+	log.Debugw("deploy started", "payload", deployPayload)
+	if scope.Bus() != nil {
+		scope.Bus().PublishID(scope.Context(), event.ETDeployStart, ref, deployPayload)
+	}
+	rollback := true
+	defer func(dp *event.DeployEvent) {
+		if rollback {
+			rp := &RemoveParams{
+				Ref: ref,
+				Revision: &dsref.Rev{
+					Field: "ds",
+					Gen:   1,
+				},
+			}
+			_, err := datasetImpl{}.Remove(scope, rp)
+			if err != nil {
+				log.Debugw("deploy rollback", "error", err)
+			}
+		}
+
+	}(&deployPayload)
+
+	go func(dp *event.DeployEvent) {
+		if err := scope.Bus().PublishID(scope.Context(), event.ETDeploySaveDatasetStart, ref, deployPayload); err != nil {
```

Comment:
any reason we're publishing pointers instead of `DeployEvent` literals? I see no situation where the event receiver should be able to modify the value. It'll have the added bonus of not needing to allocate a pointer to pass to each goroutine.

Reply:
ah, no reason, this is just a weird hold over that I need to fix. When I was publishing events in the `defer` func, I needed to pass in a pointer to the defer func, and I just had it in my head that that was necessary each time.